### DISCUSSION TITLE
Twister and sysbuild interactions (be sure to use --west-flash="--erase" when it is fixed)

### DISCUSSION THREAD

**Discussion by PerMac** [zephyrproject-rtos/zephyr#70547] [Ideas]
We were recently debugging errors that twister was reporting when running sysbuild tests. I think it is worth to share our findings in case others would spend their time trying to figure out what's wrong.

TL;DR: When testing sysbuild make sure your board is fully erased. Currently, `--west-flash=--erase` is not working correctly with sysbuild, but when it is fixed, this flag should be used in twister. https://github.com/zephyrproject-rtos/zephyr/issues/50421

Found issues without fully erased board:
Sysbuild performs multiple flashes. Due to a missing feature, each flash is followed with a reset, which starts an app hence starts printing output and twister starts reading the output looking for matching lines (either ztest frames or regexes from yaml). This can lead to a false positives, when twister recognize a failure of a previously flashed app. E.g. test nr 1 has a bootloader and a main app. The main app is a failing ztest, printing "PROJECT EXECUTION FAILED". Twister recognize it correctly and test 1 fails. Test 2 has a bootloader and another app (working correctly). After flashing the bootloader (and before app 2 is flashed) an output is printed, twister recognize "PROJECT EXECUTION FAILED", and fails the test, even though the app 2 is working, just not yet flashed. Such behavior wouldn't take place if the board was erased between tests. Due to #50421 it is not yet possible to easily automate it with twister, but a feature to properly recognized when to do erase/reset/recover etc. in sysbuild workflow is being worked on https://github.com/zephyrproject-rtos/zephyr/pull/69748. 

Also, sysbuild doesn't pass run_id to other than main app (@nordicjm a bug/missing feature?). If a board was not erased and sth went wrong (e.g. with flashing) it can happen that a main app will be the proper one and run_id will match, but some other image can be wrong and this won't be detected.

FYI @nashif @hakehuang @golowanow @gchwier @nordicjm @tejlmand

**Comment by nordicjm** [comment]
> Also, sysbuild doesn't pass run_id to other than main app (@nordicjm a bug/missing feature?). If a board was not erased and sth went wrong (e.g. with flashing) it can happen that a main app will be the proper one and run_id will match, but some other image can be wrong and this won't be detected.

With how the runid code works, it is not possible to pass this on because all images would need have ZTEST enabled, when sysbuild is used it's because of wanting multiple images e.g. bootloader and test application, so for the bootloader ztest can't be enabled and for any applications that do not use ztest (many samples/tests do not), so if this feature is wanted then it needs to be redesigned so that it can be enabled for any application type, not just those that enable ZTEST (and should use a Kconfig value for the runid instead of a cmake argument)

**Comment by hakehuang** [comment]
1. the erase issue is more complexed, e.g for  some NXP platform we need add below to erase

```
#exec EnableEraseAllFlashBanks
#erase 0x08000000, 0x085fffff
```
and as the application may start some time we need erase several times. not sure whether you have such issue at your platform? and to workaround this, I add a pre-script to do the erase.
2. besides that, the sysbuild will introduce the path in runner.yaml, which will cause --skip-build option not working if we deploy it to another machine with different code directly.
